# Chicago Neighborhood Energy Use Analysis

## Project Overview

This project analyzes the Chicago Neighborhood Energy Use dataset to produce a clean, reliable dataset suitable for exploratory data analysis (EDA). The notebook follows a structured, reproducible workflow that emphasizes evidence-based data cleaning and clear documentation. The cleaned dataset will be used to explore patterns in neighborhood energy consumption and identify meaningful insights.

## Imports and Configuration

Import the libraries required for data loading, analysis, and visualization. Configure display settings at the beginning of the notebook to improve readability and ensure consistent output throughout the analysis.

In [67]:
import numpy as np                  # Linear algebra
import pandas as pd                 # Data processing, CSV file I/O (e.g., pd.read_csv)
import matplotlib.pyplot as plt     # Matplotlib for plotting
import seaborn as sns               # Seaborn for statistical visualization

# Configure pandas display options
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:.2f}".format)

## Load the Dataset

Load the raw dataset into a pandas DataFrame using a relative file path. Using relative paths improves portability, allowing the project to be run on different systems without modifying the code.


In [68]:
# Load the dataset
df_raw = pd.read_csv("../data/raw/NN_Energy_Use_20260718.csv")
df = df_raw.copy()

print("Dataset loaded successfully.")

Dataset loaded successfully.


## Initial Inspection

The purpose of this section is to gain an initial understanding of the dataset before making any modifications. This includes reviewing the dataset's structure, dimensions, column names, data types, and summary statistics to identify areas that may require further investigation during the data cleaning process.

In [69]:
df.head()

,COMMUNITY AREA NAME,CENSUS BLOCK,BUILDING TYPE,BUILDING_SUBTYPE,KWH JANUARY 2010,KWH FEBRUARY 2010,KWH MARCH 2010,KWH APRIL 2010,KWH MAY 2010,KWH JUNE 2010,KWH JULY 2010,KWH AUGUST 2010,KWH SEPTEMBER 2010,KWH OCTOBER 2010,KWH NOVEMBER 2010,KWH DECEMBER 2010,TOTAL KWH,ELECTRICITY ACCOUNTS,ZERO KWH ACCOUNTS,THERM JANUARY 2010,THERM FEBRUARY 2010,THERM MARCH 2010,TERM APRIL 2010,THERM MAY 2010,THERM JUNE 2010,THERM JULY 2010,THERM AUGUST 2010,THERM SEPTEMBER 2010,THERM OCTOBER 2010,THERM NOVEMBER 2010,THERM DECEMBER 2010,TOTAL THERMS,GAS ACCOUNTS,KWH TOTAL SQFT,THERMS TOTAL SQFT,KWH MEAN 2010,KWH STANDARD DEVIATION 2010,KWH MINIMUM 2010,KWH 1ST QUARTILE 2010,KWH 2ND QUARTILE 2010,KWH 3RD QUARTILE 2010,KWH MAXIMUM 2010,KWH SQFT MEAN 2010,KWH SQFT STANDARD DEVIATION 2010,KWH SQFT MINIMUM 2010,KWH SQFT 1ST QUARTILE 2010,KWH SQFT 2ND QUARTILE 2010,KWH SQFT 3RD QUARTILE 2010,KWH SQFT MAXIMUM 2010,THERM MEAN 2010,THERM STANDARD DEVIATION 2010,THERM MINIMUM 2010,THERM 1ST QUARTILE 2010,THERM 2ND QUARTILE 2010,THERM 3RD QUARTILE 2010,THERM MAXIMUM 2010,THERMS SQFT MEAN 2010,THERMS SQFT STANDARD DEVIATION 2010,THERMS SQFT MINIMUM 2010,THERMS SQFT 1ST QUARTILE 2010,THERMS SQFT 2ND QUARTILE 2010,THERMS SQFT 3RD QUARTILE 2010,THERMS SQFT MAXIMUM 2010,TOTAL POPULATION,TOTAL UNITS,AVERAGE STORIES,AVERAGE BUILDING AGE,AVERAGE HOUSESIZE,OCCUPIED UNITS,OCCUPIED UNITS PERCENTAGE,RENTER-OCCUPIED HOUSING UNITS,RENTER-OCCUPIED HOUSING PERCENTAGE,OCCUPIED HOUSING UNITS
0,Near North Side,170310818002021.00,Commercial,Commercial,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,"1,744","1,511","1,164",800,554,334,296,259,299,397,571,"1,230","9,159",5,NaN,"22,100",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"9,159",NaN,"9,159","9,159","9,159","9,159","9,159","22,100",NaN,"22,100","22,100","22,100","22,100","22,100",0,0,6.00,119.00,0.00,0,NaN,0,NaN,0
1,Near North Side,170310801004004.00,Commercial,Commercial,"715,681","617,436","394,468","174,066","247,050","277,598","275,899","281,530","161,036","217,122","632,337","948,260","4,942,483",108,8,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"317,765",NaN,"4,942,483",NaN,"4,942,483","4,942,483","4,942,483","4,942,483","4,942,483","317,765",NaN,"317,765","317,765","317,765","317,765","317,765",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,158,112,1.00,0.00,1.76,90,0.80,10,0.11,90
2,Near North Side,170310804001003.00,Residential,Multi < 7,"1,959","2,039","1,647","1,504","1,790","3,340","4,368","4,376","2,426","3,208","2,236","3,536","32,429",4,2,"1,016",955,608,349,170,78,36,59,35,50,205,740,"4,301",4,"5,804","5,804","32,429",NaN,"32,429","32,429","32,429","32,429","32,429","5,804",NaN,"5,804","5,804","5,804","5,804","5,804","4,301",NaN,"4,301","4,301","4,301","4,301","4,301","5,804",NaN,"5,804","5,804","5,804","5,804","5,804",36,18,3.00,116.00,2.00,18,1.00,12,0.67,18
3,Near North Side,170310810005005.00,Commercial,Commercial,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2,"2,076","2,296","1,611",738,366,195,69,61,118,583,"1,044","2,040","11,197",Less than 4,NaN,"12,259",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"11,197",NaN,"11,197","11,197","11,197","11,197","11,197","12,259",NaN,"12,259","12,259","12,259","12,259","12,259",299,238,1.00,0.00,1.43,209,0.88,201,0.96,209
4,Near North Side,170310815002006.00,Commercial,Commercial,"1,351,588","1,229,157","943,203","864,062","954,775","985,083","1,126,242","1,088,258","986,930","924,947","1,092,475","1,577,138","13,123,858",27,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"494,300",NaN,"13,123,858",NaN,"13,123,858","13,123,858","13,123,858","13,123,858","13,123,858","494,300",NaN,"494,300","494,300","494,300","494,300","494,300",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,0,40.00,21.00,0.00,0,NaN,0,NaN,0


In [70]:
df.shape

(666, 73)

In [71]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 666 entries, 0 to 665
Data columns (total 73 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   COMMUNITY AREA NAME                  666 non-null    str    
 1   CENSUS BLOCK                         665 non-null    float64
 2   BUILDING TYPE                        665 non-null    str    
 3   BUILDING_SUBTYPE                     665 non-null    str    
 4   KWH JANUARY 2010                     635 non-null    str    
 5   KWH FEBRUARY 2010                    635 non-null    str    
 6   KWH MARCH 2010                       635 non-null    str    
 7   KWH APRIL 2010                       635 non-null    str    
 8   KWH MAY 2010                         635 non-null    str    
 9   KWH JUNE 2010                        635 non-null    str    
 10  KWH JULY 2010                        635 non-null    str    
 11  KWH AUGUST 2010                      635 no

In [72]:
df.describe()

,CENSUS BLOCK,AVERAGE STORIES,AVERAGE BUILDING AGE,AVERAGE HOUSESIZE,OCCUPIED UNITS PERCENTAGE,RENTER-OCCUPIED HOUSING PERCENTAGE
count,665.00,666.00,666.00,666.00,561.00,558.00
mean,170311509962382.59,6.83,48.00,2.11,0.86,0.49
std,2196078283.62,10.40,44.67,15.80,0.15,0.31
min,170310713001002.00,1.00,0.00,0.00,0.00,0.00
25%,170310803004000.00,1.50,1.33,1.30,0.82,0.26
50%,170310812021010.00,3.00,38.50,1.58,0.90,0.42
75%,170310817002029.00,6.00,85.00,1.92,0.95,0.77
max,170318422001079.00,74.00,134.00,408.76,1.00,1.00


### Initial Inspection Summary

The initial inspection provided a high-level overview of the dataset before any cleaning was performed.

**Key observations:**

- The dataset contains **666 observations** and **73 variables**.
- The dataset includes both **categorical** and **numeric** information, such as:
  - Monthly electricity usage in **kWh**
  - Monthly natural gas usage in **Therms**
- Several variables contain **missing values**, so data completeness will need further investigation during cleaning.
- Some monthly energy usage variables were imported as **string** data types even though they appear to contain numeric values. These columns will be reviewed before any type conversions are made.
- The descriptive statistics show values that may require further investigation, including unusually large maximum values relative to the overall distribution.
- No changes have been made to the dataset at this stage.

This section was used only to understand the dataset and identify potential data quality issues.

The next phase focuses on systematically investigating and addressing data quality issues while preserving the integrity of the original data.


## Data Cleaning

The objective of this phase is to improve the quality of the dataset while preserving its integrity. Rather than making assumptions about the data, each potential issue will be investigated before any transformation is applied.

Throughout this section, we will follow the principle:

- Investigate before you transform.

Each cleaning decision will be based on evidence gathered during the investigation rather than assumptions about the data.

### Investigate Missing Values

Before handling missing values, it is important to understand where they occur and how prevalent they are throughout the dataset. This investigation helps determine whether missing values represent isolated cases or broader data quality issues and informs the most appropriate cleaning strategy.

In [73]:
missing_summary = (
    pd.DataFrame({
        "Missing Count": df.isna().sum(),
        "Missing Percentage": (df.isna().mean() * 100)
    })
    .query("`Missing Count` > 0")
    .sort_values(by="Missing Count", ascending=False)
)

missing_summary

,Missing Count,Missing Percentage
THERMS SQFT STANDARD DEVIATION 2010,261,39.19
KWH SQFT STANDARD DEVIATION 2010,255,38.29
THERM STANDARD DEVIATION 2010,221,33.18
KWH STANDARD DEVIATION 2010,212,31.83
RENTER-OCCUPIED HOUSING PERCENTAGE,108,16.22
...,...,...
KWH 1ST QUARTILE 2010,31,4.65
KWH MEAN 2010,31,4.65
BUILDING_SUBTYPE,1,0.15
CENSUS BLOCK,1,0.15


### Investigate Missing Value Patterns

Understanding the number and percentage of missing values is only the first step. Before deciding whether any action is required, it is important to investigate how the missing values are distributed throughout the dataset. Examining missing value patterns can help determine whether the missing values share a common cause or appear to occur independently.

In [74]:
missing_rows = df[df.isna().any(axis=1)]

missing_rows

,COMMUNITY AREA NAME,CENSUS BLOCK,BUILDING TYPE,BUILDING_SUBTYPE,KWH JANUARY 2010,KWH FEBRUARY 2010,KWH MARCH 2010,KWH APRIL 2010,KWH MAY 2010,KWH JUNE 2010,KWH JULY 2010,KWH AUGUST 2010,KWH SEPTEMBER 2010,KWH OCTOBER 2010,KWH NOVEMBER 2010,KWH DECEMBER 2010,TOTAL KWH,ELECTRICITY ACCOUNTS,ZERO KWH ACCOUNTS,THERM JANUARY 2010,THERM FEBRUARY 2010,THERM MARCH 2010,TERM APRIL 2010,THERM MAY 2010,THERM JUNE 2010,THERM JULY 2010,THERM AUGUST 2010,THERM SEPTEMBER 2010,THERM OCTOBER 2010,THERM NOVEMBER 2010,THERM DECEMBER 2010,TOTAL THERMS,GAS ACCOUNTS,KWH TOTAL SQFT,THERMS TOTAL SQFT,KWH MEAN 2010,KWH STANDARD DEVIATION 2010,KWH MINIMUM 2010,KWH 1ST QUARTILE 2010,KWH 2ND QUARTILE 2010,KWH 3RD QUARTILE 2010,KWH MAXIMUM 2010,KWH SQFT MEAN 2010,KWH SQFT STANDARD DEVIATION 2010,KWH SQFT MINIMUM 2010,KWH SQFT 1ST QUARTILE 2010,KWH SQFT 2ND QUARTILE 2010,KWH SQFT 3RD QUARTILE 2010,KWH SQFT MAXIMUM 2010,THERM MEAN 2010,THERM STANDARD DEVIATION 2010,THERM MINIMUM 2010,THERM 1ST QUARTILE 2010,THERM 2ND QUARTILE 2010,THERM 3RD QUARTILE 2010,THERM MAXIMUM 2010,THERMS SQFT MEAN 2010,THERMS SQFT STANDARD DEVIATION 2010,THERMS SQFT MINIMUM 2010,THERMS SQFT 1ST QUARTILE 2010,THERMS SQFT 2ND QUARTILE 2010,THERMS SQFT 3RD QUARTILE 2010,THERMS SQFT MAXIMUM 2010,TOTAL POPULATION,TOTAL UNITS,AVERAGE STORIES,AVERAGE BUILDING AGE,AVERAGE HOUSESIZE,OCCUPIED UNITS,OCCUPIED UNITS PERCENTAGE,RENTER-OCCUPIED HOUSING UNITS,RENTER-OCCUPIED HOUSING PERCENTAGE,OCCUPIED HOUSING UNITS
0,Near North Side,170310818002021.00,Commercial,Commercial,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,"1,744","1,511","1,164",800,554,334,296,259,299,397,571,"1,230","9,159",5,NaN,"22,100",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"9,159",NaN,"9,159","9,159","9,159","9,159","9,159","22,100",NaN,"22,100","22,100","22,100","22,100","22,100",0,0,6.00,119.00,0.00,0,NaN,0,NaN,0
1,Near North Side,170310801004004.00,Commercial,Commercial,"715,681","617,436","394,468","174,066","247,050","277,598","275,899","281,530","161,036","217,122","632,337","948,260","4,942,483",108,8,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"317,765",NaN,"4,942,483",NaN,"4,942,483","4,942,483","4,942,483","4,942,483","4,942,483","317,765",NaN,"317,765","317,765","317,765","317,765","317,765",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,158,112,1.00,0.00,1.76,90,0.80,10,0.11,90
2,Near North Side,170310804001003.00,Residential,Multi < 7,"1,959","2,039","1,647","1,504","1,790","3,340","4,368","4,376","2,426","3,208","2,236","3,536","32,429",4,2,"1,016",955,608,349,170,78,36,59,35,50,205,740,"4,301",4,"5,804","5,804","32,429",NaN,"32,429","32,429","32,429","32,429","32,429","5,804",NaN,"5,804","5,804","5,804","5,804","5,804","4,301",NaN,"4,301","4,301","4,301","4,301","4,301","5,804",NaN,"5,804","5,804","5,804","5,804","5,804",36,18,3.00,116.00,2.00,18,1.00,12,0.67,18
3,Near North Side,170310810005005.00,Commercial,Commercial,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2,"2,076","2,296","1,611",738,366,195,69,61,118,583,"1,044","2,040","11,197",Less than 4,NaN,"12,259",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"11,197",NaN,"11,197","11,197","11,197","11,197","11,197","12,259",NaN,"12,259","12,259","12,259","12,259","12,259",299,238,1.00,0.00,1.43,209,0.88,201,0.96,209
4,Near North Side,170310815002006.00,Commercial,Commercial,"1,351,588","1,229,157","943,203","864,062","954,775","985,083","1,126,242","1,088,258","986,930","924,947","1,092,475","1,577,138","13,123,858",27,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"494,300",NaN,"13,123,858",NaN,"13,123,858","13,123,858","13,123,858","13,123,858","13,123,858","494,300",NaN,"494,300","494,300","494,300","494,300","494,300",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,0,40.00,21.00,0.00,0,NaN,0,NaN,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,

### Missing Values Investigation Summary

The investigation identified missing values in several variables throughout the dataset. Missing value counts and percentages were calculated to assess the extent of missingness, and observations containing missing values were inspected to identify broad patterns.

Key findings include:

- Several variables contain missing values, with the highest percentages occurring in variables related to standard deviation measurements.
- No variable consists entirely of missing values.
- Observations containing missing values are distributed across the dataset rather than being isolated to a small subset of records.
- Missing values in the monthly electricity (kWh) and natural gas (Therms) variables appear to occur together for certain observations, suggesting a structured pattern of missingness that warrants further investigation.
- At this stage, no transformations have been applied. Additional investigation will determine the appropriate treatment for missing values as other aspects of data quality are evaluated.

### Investigate Duplicate Observations

Duplicate observations can affect summary statistics, visualizations, and analytical conclusions. Before removing any duplicates, it is important to determine whether they represent data entry errors, repeated records, or legitimate observations.

The purpose of this investigation is to identify duplicate observations and determine whether any action is necessary. At this stage, no duplicate observations will be removed.

In [75]:
duplicate_rows = df.duplicated()

duplicate_summary = pd.DataFrame({
    "Metric": [
        "Duplicate Observations",
        "Percentage of Dataset"
    ],
    "Value": [
        duplicate_rows.sum(),
        duplicate_rows.mean() * 100
    ]
})

duplicate_summary

,Metric,Value
0,Duplicate Observations,0.00
1,Percentage of Dataset,0.00


### Duplicate Observations Summary

The dataset was examined for exact duplicate observations using all variables. No duplicate observations were identified.

Since no exact duplicates are present, no action is required. The original dataset remains unchanged, and the investigation will proceed to the next aspect of data quality.

### Investigate Data Types

Data types determine how variables are stored, interpreted, and analyzed. Incorrect data types can prevent calculations, produce misleading results, or limit the types of analyses that can be performed.

The purpose of this investigation is to identify variables whose data types may not accurately represent the underlying data. Any potential issues will be investigated before considering data type conversions.

In [76]:
df. dtypes.value_counts()

str        67
float64     6
Name: count, dtype: int64

#### Investigate Non-Numeric Measurement Values

In [77]:
df.rename(columns={"TERM APRIL 2010": "THERM APRIL 2010"}, inplace=True)

string_columns = df.select_dtypes(include="string").columns

string_columns

Index(['COMMUNITY AREA NAME', 'BUILDING TYPE', 'BUILDING_SUBTYPE',
       'KWH JANUARY 2010', 'KWH FEBRUARY 2010', 'KWH MARCH 2010',
       'KWH APRIL 2010', 'KWH MAY 2010', 'KWH JUNE 2010', 'KWH JULY 2010',
       'KWH AUGUST 2010', 'KWH SEPTEMBER 2010', 'KWH OCTOBER 2010',
       'KWH NOVEMBER 2010', 'KWH DECEMBER 2010', 'TOTAL KWH',
       'ELECTRICITY ACCOUNTS', 'ZERO KWH ACCOUNTS', 'THERM JANUARY 2010',
       'THERM FEBRUARY 2010', 'THERM MARCH 2010', 'THERM APRIL 2010',
       'THERM MAY 2010', 'THERM JUNE 2010', 'THERM JULY 2010',
       'THERM AUGUST 2010', 'THERM SEPTEMBER 2010', 'THERM OCTOBER 2010',
       'THERM NOVEMBER 2010', 'THERM DECEMBER 2010', 'TOTAL THERMS',
       'GAS ACCOUNTS', 'KWH TOTAL SQFT', 'THERMS TOTAL SQFT', 'KWH MEAN 2010',
       'KWH STANDARD DEVIATION 2010', 'KWH MINIMUM 2010',
       'KWH 1ST QUARTILE 2010', 'KWH 2ND QUARTILE 2010',
       'KWH 3RD QUARTILE 2010', 'KWH MAXIMUM 2010', 'KWH SQFT MEAN 2010',
       'KWH SQFT STANDARD DEVIATION 2010', '

In [78]:
energy_columns = [
    column
    for column in string_columns
    if "KWH" in column or "THERM" in column
]

energy_columns

57


In [79]:
deferred_conversion_columns = [
    "ELECTRICITY ACCOUNTS",
    "GAS ACCOUNTS",
]

In [80]:
validated_numeric_columns = [
    column
    for column in string_columns
    if column not in energy_columns
       and column not in deferred_conversion_columns
       and "COMMUNITY" not in column
       and "BUILDING" not in column
]

validated_numeric_columns

['TOTAL POPULATION',
 'TOTAL UNITS',
 'OCCUPIED UNITS',
 'RENTER-OCCUPIED HOUSING UNITS',
 'OCCUPIED HOUSING UNITS']

In [81]:
numeric_candidate_columns = (
        energy_columns + validated_numeric_columns
)

numeric_candidate_columns

62


In [82]:
invalid_numeric_values = {}

for column in numeric_candidate_columns:
    cleaned_values = (
        df[column]
        .dropna()
        .str.replace(",", "", regex=False)
    )

    invalid_values = cleaned_values[
        pd.to_numeric(cleaned_values, errors="coerce").isna()
    ].unique()

    if len(invalid_values) > 0:
        invalid_numeric_values[column] = invalid_values

if invalid_numeric_values:
    print("Columns containing unexpected non-numeric values:\n")

    for column, values in invalid_numeric_values.items():
        print(f"{column}: {values}\n")
else:
    print("Validation successful: "
          "All candidate numeric columns contain valid numeric values."
          )

Validation successful: All candidate numeric columns contain valid numeric values.


#### Non-Numeric Measurement Values Summary

Validation identified the value "Less than 4" in the **ELECTRICITY ACCOUNTS** and **GAS ACCOUNTS** variables. These columns otherwise contain numeric values and represent quantitative account counts.

The available dataset documentation confirms that both variables represent utility accounts counts but does not explain the meaning or intended treatment of the value "Less than 4". Because its exact numeric value cannot be determined from the available documentation, no transformation will be applied to these columns at this stage.

These variables have been separated into a dedicated list of deferred conversion columns and will not be converted during this phase of the cleaning process. This decision preserves the integrity of the original data and avoids introducing unsupported assumptions into the analysis. As a result, these variables remain stored as text for the remainder of the cleaning process. This reflects a temporary implementation decision rather than a change in their intended analytical meaning as quantitative variables.

#### Convert Validated Numeric Columns

The candidate numeric columns were validated before conversion to ensure that they contained only numeric values (after removing formatting characters such as thousands separators). Two variables (`ELECTRICITY ACCOUNTS` and `GAS ACCOUNTS`) contained the undocumented value **"Less than 4"** and were excluded from conversion pending additional authoritative guidance.

All remaining validated numeric columns are converted to numeric data types in the following step using `errors="raise"` to ensure that any unexpected values are reported immediately.


In [83]:
for column in numeric_candidate_columns:
    df[column] = pd.to_numeric(
        df[column].str.replace(",", "", regex=False),
        errors="raise"
    )

In [84]:
df[numeric_candidate_columns].dtypes.value_counts()

float64    56
int64       6
Name: count, dtype: int64

In [85]:
df[deferred_conversion_columns].dtypes

ELECTRICITY ACCOUNTS    str
GAS ACCOUNTS            str
dtype: object

#### Investigate Identifier Variables

The `CENSUS BLOCK` variable is an identifier rather than a quantitative measurement. Although pandas imported the column as `float64`, this was caused by the presence of a missing value, which required the use of a floating-point data type.

The online dataset stores census block values without decimal places, confirming that the decimal portion introduced by pandas is a storage artifact rather than part of the original data. To better reflect the semantic meaning of the variable, `CENSUS BLOCK` will be converted to a string identifier while preserving missing values.

In [94]:
df["CENSUS BLOCK"] = (
    df["CENSUS BLOCK"]
    .astype("Int64")
    .astype("string")
)

In [96]:
categorical_text_columns = [
    "COMMUNITY AREA NAME",
    "BUILDING TYPE",
    "BUILDING_SUBTYPE",
]

for column in categorical_text_columns:
    df[column] = df[column].astype("string")

In [100]:
df[["CENSUS BLOCK"] + categorical_text_columns].dtypes

CENSUS BLOCK           string
COMMUNITY AREA NAME    string
BUILDING TYPE          string
BUILDING_SUBTYPE       string
dtype: object

In [103]:
df.dtypes.value_counts()

float64    61
int64       6
string      4
str         2
Name: count, dtype: int64